bolt/test/binary-analysis/AArch64/gs-pacret-autiasp.s

```diff
@@ -0,0 +1,914 @@
+
+        .globl  f_nonx30_ret_ok
+        .type   f_nonx30_ret_ok,@function
+f_nonx30_ret_ok:
+        paciasp
+        stp     x29, x30, [sp, #-16]!
+        mov     x29, sp
+        bl      g
+        add     x0, x0, #3
+        ldp     x29, x30, [sp], #16
+        // FIXME: Should the scanner understand that an authenticated register (below x30,
+        //        after the autiasp instruction), is OK to be moved to another register
+        //        and then that register being used to return?
+        //        This respects that pac-ret hardening intent, but the scanner currently
+        //        will produce a false positive for this.
+        //        Is it worthwhile to make the scanner more complex for this case?
+        //        So far, scanning many millions of instructions across a linux distro,
+        //        I haven't encountered such an example.
+        //        The ".if 0" block below tests this case and currently fails.
+.if 0
+        autiasp
+        mov     x16, x30
+.else
+        mov     x16, x30
+        autia   x16, sp
+.endif
+// CHECK-NOT: function f_nonx30_ret_ok
+        ret     x16
+        .size f_nonx30_ret_ok, .-f_nonx30_ret_ok
+
+
+        .globl  f_callclobbered_x30
+        .type   f_callclobbered_x30,@function
+f_callclobbered_x30:
+        bl      g
+// CHECK-LABEL: GS-PACRET: non-protected ret found in function f_callclobbered_x30, basic block .LBB{{[0-9]+}}, at address
+// CHECK-NEXT:    The return instruction is     {{[0-9a-f]+}}:       ret
+// CHECK-NEXT:    The 1 instructions that write to the return register after any authentication are:
+// CHECK-NEXT:    1. {{[0-9a-f]+}}: bl      g@PLT
+// CHECK-NEXT:  This happens in the following basic block:
+// CHECK-NEXT: {{[0-9a-f]+}}:   bl      g@PLT
+// CHECK-NEXT: {{[0-9a-f]+}}:   ret
+        ret
+        .size f_callclobbered_x30, .-f_callclobbered_x30
+
+        .globl  f_callclobbered_calleesaved
+        .type   f_callclobbered_calleesaved,@function
+f_callclobbered_calleesaved:
+        bl      g
+// CHECK-LABEL: GS-PACRET: non-protected ret found in function f_callclobbered_calleesaved, basic block .LBB{{[0-9]+}}, at address
+// CHECK-NEXT:    The return instruction is     {{[0-9a-f]+}}:       ret x19
+// CHECK-NEXT:    The 1 instructions that write to the return register after any authentication are:
+// CHECK-NEXT:    1. {{[0-9a-f]+}}: bl      g@PLT
+// CHECK-NEXT:  This happens in the following basic block:
+// CHECK-NEXT: {{[0-9a-f]+}}:   bl      g@PLT
+// CHECK-NEXT: {{[0-9a-f]+}}:   ret x19
+        // x19, according to the Arm ABI (AAPCS) is a callee-saved register.
+        // Therefore, if function g respects the AAPCS, it should not write
+        // anything to x19. However, we can't know whether function g actually
+        // does respect the AAPCS rules, so the scanner should assume x19 can
+        // get overwritten, and report a gadget if the code does not properly
+        // deal with that.
```

Comment:
Not only that, but also if it's callee-saved then there's a good chance that it's been saved on a stack at some point, so this is equivalent to having been loaded from memory even the callee does attempt to restore it.

Reply:
Ah, yes, excellent point, thanks. I've improved the comment both here in the test case and in the analysis implementation to try and cover that.